### PR TITLE
Move features from passOptions to Module

### PR DESCRIPTION
**Pull request by tlively**
This allowing us to emit a (potentially modified) target features
section and conditionally emit other sections such as the DataCount
section based on the presence of features.

### CODE REVIEW

**Comment by tlively** [comment]
This is essentially https://github.com/WebAssembly/binaryen/pull/1993 without the controversial parts.

File: src/tools/tool-options.h [hunk 1]
```diff
@@ -91,36 +83,32 @@ struct ToolOptions : public Options {
...
              });
     return *this;
   }
 
-  void calculateFeatures(const Module& module) {
-    FeatureSet wasmFeatures;
-    bool wasmHasFeatures =
-        ModuleUtils::readFeaturesSection(module, wasmFeatures);
-
+  void calculateFeatures(Module& module) {
```
Comment (kripken) [inline src/tools/tool-options.h]: how about "applyFeatures"? I think that makes it clear it applies them to the module.

File: src/wasm.h [hunk 2]
```diff
@@ -924,6 +924,11 @@ class Module {
   std::vector<UserSection> userSections;
   std::vector<std::string> debugInfoFileNames;
 
+  // `features` are the features allowed to be used in this module and should be
+  // respected regardless of the value of `hasFeaturesSection`.
```
Comment (kripken) [inline src/wasm.h]: maybe also add that `hasFeaturesSection` means that we read one, and will write one (unless told not to)

File: src/wasm/wasm-binary.cpp [hunk 3]
```diff
@@ -666,8 +706,8 @@ void WasmBinaryBuilder::readUserSection(size_t payloadLen) {
 
 uint8_t WasmBinaryBuilder::getInt8() {
   if (!more()) throwError("unexpected end of input");
-  if (debug) std::cerr << "getInt8: " << (int)(uint8_t)input[pos] << " (at " << pos << ")" << std::endl;
-  return input[pos++];
+  if (debug) std::cerr << "getInt8: " << (int)(uint8_t)input.at(pos) << " (at " << pos << ")" << std::endl;
+  return input.at(pos++);
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: were these changes intended? i don't think we need them.

File: scripts/test/shared.py [hunk 4]
```diff
@@ -419,12 +416,10 @@ def minify_check(wast, verify_final_result=True):
...
   cmd = WASM_OPT + [wast, '--print-minified']
   print '      ', ' '.join(cmd)
-  subprocess.check_call(
-      WASM_OPT + [wast, '--print-minified'],
-      stdout=open('a.wast', 'w'), stderr=subprocess.PIPE)
+  subprocess.check_call(cmd, stdout=open('a.wast', 'w'), stderr=subprocess.PIPE)
   assert os.path.exists('a.wast')
   subprocess.check_call(
-      WASM_OPT + ['a.wast', '--print-minified'],
+      WASM_OPT + ['a.wast', '--print-minified', '-all'],
```
[the remark this answers is not in the document]
Comment (tlively) [reply scripts/test/shared.py]: done.

File: src/wasm/wasm-binary.cpp (hunk 3 above)
Comment (tlively) [reply src/wasm/wasm-binary.cpp]: true, reverted.

File: src/wasm.h (hunk 2 above)
Comment (tlively) [reply src/wasm.h]: done.

File: src/tools/tool-options.h (hunk 1 above)
Comment (tlively) [reply src/tools/tool-options.h]: done.

**Comment by tlively** [comment]
.map files changed because I changed FeatureSet::iterFeatures to match the alphabetical order of feature names (using the better feature names used by LLVM, which we should eventually switch to in Binaryen as well).

**Comment by kripken** [comment]
Sounds good, lgtm.

**Comment by tlively** [comment]
@kripken I just want to make sure you don't have an issue with the last two commits before merging. Apparently one of the spec tests expects an exported mutable global to fail validation, so I had to disable mutable globals for the spect tests and that meant adding feature options to wasm-shell. I'm surprised the spec test does not expect mutable globals to be enabled.

**Comment by kripken** [comment]
The spec test is probably just quite old. How about simply removing the silly spec test?

**Comment by kripken** [comment]
(Eventually it would be nice if the spec test files specified which features they use, or something like that, maybe...)

**Comment by sbc100** [comment]
Somehow this broke at least the other.test_zeroinit test.  Seems that zeros are no longer being compressed in memory segments?  Does that make any sense?

File: src/ir/memory-utils.h [hunk 5]
```diff
@@ -69,6 +69,12 @@ namespace MemoryUtils {
       return true;
     }
 
+    // Conservatively refuse to change segments if there might be memory.init
+    // and data.drop instructions.
```
Comment (sbc100) [inline src/ir/memory-utils.h]: This looks like the problem

**Comment by tlively** [comment]
Yes, --detect-features is not properly defaulting to MVP in the absence of a features section. I'm working on a fix now.
